packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -117,28 +125,37 @@ const FreehandLineModel = types
   }))
   .actions((self) => ({
     initialPosition({ x, y }) {
-      self.points.push({ x, y })
+      self.x.push(x)
+      self.y.push(y)
       self.dragPoint = null
       self.targetPoint = null
       self.clipPath = []
     },
 
     initialDrag({ x, y }) {
-      self.points.push({ x: x, y: y })
+      const dist = self.getDistance(x, y, self.lastPoint.x, self.lastPoint.y)
+      if (dist > LINE_RESOLUTION) {
+        self.x.push(x)
+        self.y.push(y)
+      }
     },
 
     setCoordinates(points) {
-      self.points = points.map(({ x, y }) => ({ x, y }))
+      self.x = points.map(({ x, y }) => x)
+      self.y = points.map(({ x, y }) => y)
     },
 
     move() {
       return
     },
 
     appendPath({ x, y }) {
-      if (!self.isClosed) {
-        self.points.push({ x, y})
-      } else {
+      const dist = self.getDistance(x, y, self.lastPoint.x, self.lastPoint.y)
```

Comment:
I'm concerned that `self.lastPoint` could be `null` at some point, and this would cause a crash - but I haven't been able to trigger such a condition.

Reply:
You've already covered this in #3667, I think.